lib/frame.c

```diff
@@ -84,19 +84,27 @@ int quicly_decode_ack_frame(const uint8_t **src, const uint8_t *end, quicly_ack_
     for (i = 0; i != frame->num_gaps; ++i) {
         if ((tmp = quicly_decodev(src, end)) == UINT64_MAX)
             goto Error;
-        frame->gaps[i] = tmp + 1;
-        if (frame->smallest_acknowledged < frame->gaps[i])
+        curr_gap = tmp + 1;
+        if (frame->smallest_acknowledged < curr_gap)
```

Comment:
This looks wired.

`frame->smallest_acknowledged` is the minimum packet number acknowledged by the first MAX_GAPS +1 ack ranges. But we keep comparing that against gaps that follow them.

I think it might be possible to split decoding vs. validating-and-filling-ack_frame_t. Like:
```c
if ((num_gaps = quicly_decodev(src, end)) == UINT64_MAX))
    goto Exit;
...
frame->num_gaps = 0;
for (i = 0; i != num_gaps; ++i) {
    if ((gap = quicly_decodev(src, end)) == UINT64_MAX)
        goto Error;
    if ((ack_range = quicly_decodev(src, end)) == UINT64_MAX)
        goto Error;
    if (i < QUICLY_ACK_MAX_GAPS) {
        /* validate and stuff quicly_decode_ack_frame_t */
        if (frame->smallest_acknowledged < gap + ack_range + 2)
            goto Error;
        frame->gaps[i] = gap + 1;
        frame->ack_block_lengths[i + 1] = ack_range + 1;
        frame->smallest_acknowledged -= gap + ack_range + 2;
        frame->num_gaps = i + 1;
    }
}
```

Reply:
I like this organization, and how we no longer have the `tmp` var all over the place. This also eliminates the need to calibrate the `frame->num_gaps` value.

Reply:
In fact I removed the `tmp` var entirely, as it was only used for the `First ACK Range (i)` in the frame. For this purpose, we can just use the `ack_range` var. The vocabulary used in the code is now closer to the spec, which I like -- 8cede031f683d3497600b90b70f462e30b91072f